### DISCUSSION TITLE
How to send an email

### DISCUSSION THREAD

**Discussion by lamafab** [Byron/google-apis-rs#345] [Q&A] [answered]
Hello

Could you please clarify how one sends an email? Once you have the [`UserMessageSendCall`](https://docs.rs/google-gmail1/latest/google_gmail1/api/struct.UserMessageSendCall.html) type, the only options I see to continue are `upload` and `upload_resumable`. How does one just send the message without any file?

**Comment by Byron** [comment, marked as answer]
Generally I recommend going by examples in [the official documentation](https://developers.google.com/gmail/api/) and then trying to map this to the generated types and methods provided here.

This project is about generating crates based on an API description and thus can't provide any support on how to use the generated crates, beyond addressing bugs and shortcomings of course.

**Comment by jacek-kurlit** [comment]
I have decided to answer this question after struggling to send any email.
I had big problems since for me this API is not very intuitive.
Please bear in mind that I'm not sure if this is correct way but it works in my project.

```rust
use std::{fs, io::Cursor, path::PathBuf};

use google_gmail1::{
    api::Message,
    hyper::{self, client::HttpConnector},
    hyper_rustls::{self, HttpsConnector},
    oauth2::{self, authenticator_delegate::InstalledFlowDelegate},
    Gmail,
};
use lettre::message::{header, Body, MultiPart, SinglePart};

use crate::{google_api::load_secret, models::AppError};

pub async fn send_email(
    topic: &str,
    content: &str,
    gpc_path: PathBuf,
    token_store_path: PathBuf,
    flow_delegate: Box<dyn InstalledFlowDelegate>,
) -> Result<(), AppError> {
    let gmail = create_gmail_service(gpc_path, token_store_path, flow_delegate).await?;
    let pdf = fs::read("/path/to/sample.pdf")?;
    let pdf_body = Body::new(pdf);

    let email = lettre::Message::builder()
        .from("from@gmail.com".parse().unwrap())
        .to("to@gmail.com".parse().unwrap())
        .subject(topic)
        .multipart(
            MultiPart::mixed()
                .singlepart(
                    SinglePart::builder()
                        .header(header::ContentType::parse("text/plain; charset=utf8")?)
                        .body(content.to_string()),
                )
                .singlepart(
                    SinglePart::builder()
                        .header(header::ContentType::parse("application/pdf")?)
                        .header(header::ContentDisposition::attachment("output.pdf"))
                        .body(pdf_body),
                ),
        )
        .map_err(|e| AppError::Internal(e.to_string()))?;
    let response = gmail
        .users()
        .messages_send(Message::default(), "me")
        .upload_resumable(
            Cursor::new(email.formatted()),
            "message/rfc822".parse().expect("mime rfc822 is valid"),
        )
        .await?;
    println!("response: {:?}", response);

    Ok(())
}
```

Technically gmail expects data in rfc822 format which can be provided in `Message.raw` but we are still obligated by `.upload_resumable()` to give some kind of data which is misleading. Therefore I'm using `lettre` crate to create rfc822 data and feed it to this method. As result I have received email with topic, content and pdf attachment.
 I hope this will help someone

**Reply by Byron** [reply]
Thanks so much for sharing! I truly wish that Google would finally provide their own bindings - these ones have been created many years ago and were barely ever used. Thus they are certainly not as good as they could be.

**Reply by lamafab** [reply]
APIs are hard, I guess

**Reply by c-git** [reply]
I would love if Google did provide bindings but I really want to say a big thank you to @Byron  for all the work you've done. It's clear you put a lot of effort into trying to help everyone. I need to make use of the Google API's for a project I'm working on and this is currently the best one I can find. There is also another set of binding by oxide computers but they don't seem to need it anymore and don't seem to be doing updates to it, hence I decided to try to help out here instead.

**Comment by c-git** [comment]
If you find this an all you really want to do is send emails it's actually easier to use the `lettre` crate directly. 
They have a pretty [clear example](https://github.com/lettre/lettre/blob/master/examples/smtp_tls.rs) on in their crate.
I'd just add that you should ensure you have 2 factor authentication enabled and then that will allow you to create an APP password that you can use for this purpose.
You'll also need to enable POP in settings.
